store/store.go

```diff
@@ -499,6 +506,65 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
+// notifyAndAdvance will notify waiters in heightSub and advance contiguousHead
+// based on already written headers.
+func (s *Store[H]) notifyAndAdvance(ctx context.Context, headers ...H) {
+	// always inform heightSub about new headers seen
+	for _, h := range headers {
+		s.heightSub.Notify(h.Height())
+	}
+
+	currHead := s.contiguousHead.Load()
+	if currHead != nil {
+		s.advanceContiguousHead(ctx, (*currHead).Height())
+	}
+}
+
+// advanceContiguousHead return a new highest contiguous height
+// or a given if not found.
+func (s *Store[H]) advanceContiguousHead(ctx context.Context, currHeight uint64) uint64 {
+	// TODO(cristaloleg): benchmark this timeout or make it dynamic.
+	advCtx, advCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer advCancel()
+
+	prevHeight := currHeight
+	var newHead H
+	for {
+		h, err := s.getByHeight(advCtx, currHeight+1)
+		if err != nil {
+			break
+		}
+		newHead = h
+		currHeight++
+	}
+
+	if currHeight > prevHeight {
+		s.updateContiguousHead(ctx, newHead, currHeight)
+	}
+	return currHeight
+}
+
+func (s *Store[H]) updateContiguousHead(ctx context.Context, newHead H, newHeight uint64) {
```

Comment:
can't we only pass newHead and do head.Height inside?

Reply:
Nice catch, done!